File: frontend/lib/models/event.dart
class Event {
  final String id;
  final String title;
  final String? description;
  final String facilityId;
  final DateTime startDate;
  final DateTime? endDate;
  final DateTime? createdAt;
  final String status;
  final List<String> walls;

  const Event({
    required this.id,
    required this.title,
    this.description,
    required this.facilityId,
    required this.startDate,
    this.endDate,
    this.createdAt,
    this.status = 'active',
    this.walls = const [],
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic v) {
      if (v == null) return null;
      if (v is String) return DateTime.tryParse(v);
      return null;
    }

    return Event(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      title: (json['title'] ?? '').toString(),
      description: json['description']?.toString(),
      facilityId: (json['facility'] ?? '').toString(),
      startDate: parseDate(json['startDate']) ?? DateTime.now(),
      endDate: parseDate(json['endDate']),
      createdAt: parseDate(json['createdAt']),
      status: json['status']?.toString() ?? 'active',
      walls: (json['walls'] as List<dynamic>?)?.map((e) => e.toString()).toList() ?? [],
    );
  }
}

File: backend/models/Event.js
const mongoose = require("mongoose");

const eventSchema = new mongoose.Schema(
    {
        title: {
            type: String,
            required: [true, "Title is required"],
            trim: true,
            maxlength: [100, "Title cannot exceed 100 characters"],
        },
        description: {
            type: String,
            trim: true,
            maxlength: [1000, "Description cannot exceed 1000 characters"],
        },
        isGlobal: {
            type: Boolean,
            default: false,
        },
        facility: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "Facility",
            required: function() {
                return !this.isGlobal && !this.groupId;
            },
        },
        groupId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "Group",
        },
        createdBy: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true,
        },
        walls: [{
            type: mongoose.Schema.Types.ObjectId,
            ref: "IndoorWall"
        }],
        startDate: {
            type: Date,
            required: [true, "Start date is required"],
        },
        endDate: {
            type: Date,
            validate: {
                validator: function (val) {
                    return !val || val >= this.startDate;
                },
                message: "End date must be on or after start date",
            },
        },
        status: {
            type: String,
            enum: ["active", "closed"],
            default: "active",
        },
    },
    {
        timestamps: true,
        toJSON: {
            transform: (doc, ret) => {
                ret.id = ret._id;
                delete ret._id;
                delete ret.__v;
                return ret;
            },
        },
    },
);

module.exports = mongoose.model("Event", eventSchema);
